### PR TITLE
Add a counter server example, with tests compatible with Concuerror

### PR DESCRIPTION
**Pull request by aronisstav**
There are some pending tasks before this is at a satisfactory level:

- [x] ~Add some way for running all tests with Concuerror.~ (Different repo)
- [x] ~Add some way for producing a table from the results of the tests.~ (Different repo)
- [x] Add test with network failure.
- [x] Add test with node failure.

### CODE REVIEW

**Comment by dszoboszlay** [comment]
I would recommend moving the test code to a new `test` directory.

As for running the tests, I think we can lift the `run_concuerror.es` script from our KDB model over to this repo. It's not rebar3 integrated of course, but since we have a `Makefile` anyway, it wouldn't be so hard to titch them together.

File: test/counter_server_example/src/counter_server_supervisor.erl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    end,
+  supervisor:start_link(SupervisorName, ?MODULE, [Arg]).
+
+%%%===================================================================
+%%% Supervisor callbacks
+%%%===================================================================
+
+init([Arg]) ->
+
+  ?TABLE = ets:new(?TABLE, [named_table, public]),
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_supervisor.erl]: The table name shall also be vnode-local in case of using `vnet`! Something like:

```erlang
TableName =
  case Arg of
    no_vnet -> ?TABLE;
    vnet -> vnet:tab(?TABLE)
  end,
```

File: test/counter_server_example/src/counter_server_example.erl [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+  catch
+    _:_ -> error
+  end.
+
+%%%-------------------------------------------------------------------
+
+check_results(error, _Res2) -> ok;
+check_results(_Res1, error) -> ok;
+check_results({ok, _Gen1, Res1}, {ok, _Gen2, Res2}) ->
+  ?assert(Res2 > Res1);
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_example.erl]: Shouldn't `Gen2 >= Gen1` be true as well?

File: test/counter_server_example/src/counter_server_example.erl [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+  valid_client_no_vnet().
+
+%%%-------------------------------------------------------------------
+
+%% @doc Client making a second request if the first was successful and
+%% verifying that the second answer is greater than the first.
+valid_client_no_vnet() ->
+  Res1 = valid_local_request(),
+  case Res1 =:= error of
+    true -> ok;
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_example.erl]: There's a `check_results/2` branch for handling `Res1 =:= error`, so why don't you do a second request here?

File: test/counter_server_example/src/counter_server_example.erl [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+   ]).
+
+%%%-------------------------------------------------------------------
+
+-export(
+   [ setup_server/0
+   , once_valid_client/0
+   , twice_valid_client/0
+   , invalid_client/0
+   , twice_valid_same_gen_client/0
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_example.erl]: For me it's very hard to follow the purpose of the functions in this module. If I understand correctly, you have:
* EUnit test functions (`*_test/0`) that will get auto-exported by EUnit, and test the no-vnet scenarios.
* Concuerror test functions exported here that test the vnet-scenarios.
* Functions that exported so that they can be used together with `vnet:rpc/4`.
* Other helper functions.

I would prefer if these classes of functions would be coupled together in the file with appropriate headers and preferably with some common naming convention to make them easy to identify.

File: test/counter_server_example/src/counter_server_example.erl [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+valid_rpc_request() ->
+  server_request(increment).
+
+server_request(Request) ->
+  try
+    %% %% The commented out variant leads to schedulings where
+    %% %% vnet_proxy.erl, l52 crashes cause the process is not
+    %% %% alive and the call returns `undefined'
+    %% counter_server:request({via, vnet, {server, ?SERVER}}, increment)
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_example.erl]: This one is on me to fix.

Actually, I think it would be beneficial to show both `rpc`+local `gen_server:call` and remote `gen_server:call` scenarios work equally under our model network.

File: test/counter_server_example/src/counter_server_example.erl [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+check_same_gen({ok, G1, _R1}, {ok, G2, _R2}) ->
+  ?assertEqual(G1, G2);
+check_same_gen(Res1, Res2) ->
+  ?assertEqual(ok, {unexpected, Res1, Res2}).
+
+%%%===================================================================
+
+%% @doc
+%% Simulating a scenario with one node as client, making a request.
+once_valid_test() ->
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_example.erl]: I'm not sure this test adds any real value. For example you don't have non-vnet variant of this scenario either.

**Review by dszoboszlay — CHANGES_REQUESTED** [review]

File: test/counter_server_example/src/counter_server_example.erl (hunk 3 above)
Comment (aronisstav) [reply test/counter_server_example/src/counter_server_example.erl]: This is purely for the sake of reducing the number of schedulings that cannot expose errors. If the first request has failed there is nothing to compare against.

File: test/counter_server_example/src/counter_server_example.erl (hunk 4 above)
Comment (aronisstav) [reply test/counter_server_example/src/counter_server_example.erl]: The module is written as an 'essay'. It is supposed to be read from the top to the bottom, and every time sth new is introduced it is defined right underneath. The _test functions are all Concuerror tests and I am indeed taking advantage of EUnit to have them exported.

There is a comment at the top of the file making an attempt to explain that.

File: src/vnet_proxy.erl [hunk 7]
```diff
@@ -24,10 +24,19 @@
...
+    fun() ->
+        init(VNodeC, VNodeS, Pid, Gen),
+        P ! {ack, Ref},
+        loop(Pid)
+    end,
+  Proxy = spawn(Fun),
+  receive
+    {ack, Ref} -> ok
+  end,
+  {ok, Proxy}.
```
Comment (dszoboszlay) [inline src/vnet_proxy.erl]: I would slightly prefer the `proc_lib` way of using `spawn_monitor` and avoiding deadlock on errors in `init/3`. But if you go with this solution, please at least update the `-spec`, as the function can no longer return errors.

Also, unexport `init/4`.

File: test/counter_server_example/src/counter_server.erl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%%%-------------------------------------------------------------------
+
+-include("counter_server.hrl").
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server.erl]: Not needed.

File: test/counter_server_example/src/counter_server_supervisor.erl [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        ]).
+
+-behaviour(supervisor).
+-export([init/1]).
+
+%%%-------------------------------------------------------------------
+
+-include("counter_server.hrl").
+
+-define(SUPERVISOR, ?MODULE).
```
Comment (dszoboszlay) [inline test/counter_server_example/src/counter_server_supervisor.erl:22]: This is the only module using the `?TABLE` macro. For consistency, I would either move the definition of `TABLE` here, or define the `SUPERVISOR` macro in the header too.

File: test/counter_server_example/test/counter_server_example.erl [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+%%%-------------------------------------------------------------------
+
+-include("../src/counter_server.hrl").
+
+%%%-------------------------------------------------------------------
+
+%% @doc
+%% Simulating a scenario with one client, making requests and checking
+%% results, without using any vnet logic.
+valid_no_vnet_test() ->
```
Comment (dszoboszlay) [inline test/counter_server_example/test/counter_server_example.erl]: Please don't make these functions look like EUnit tests! If you run `make`, `rebar3` will try to execute them as such and fail (`{badmatch,{error,{already_started,<0.128.0>}}}`).

**Review by dszoboszlay — CHANGES_REQUESTED** [review]

File: test/counter_server_example/test/counter_server_example.erl (hunk 10 above)
Comment (aronisstav) [reply test/counter_server_example/test/counter_server_example.erl]: ok!

File: test/counter_server_example/src/counter_server_supervisor.erl (hunk 9 above)
Comment (aronisstav) [reply test/counter_server_example/src/counter_server_supervisor.erl:22]: Moved `?TABLE` here.

**Review by aronisstav — COMMENTED** [review]
All comments have been addressed.

**Comment by aronisstav** [comment]
@dszoboszlay @viktoriafordos Can we merge this?

File: src/vnet_proxy.erl [hunk 11]
```diff
@@ -24,10 +23,19 @@
...
+  Ref = make_ref(),
+  Fun =
+    fun() ->
+        init(VNodeC, VNodeS, Pid, Gen),
+        P ! {ack, Ref},
+        loop(Pid)
+    end,
+  {Proxy, Monitor} = spawn_monitor(Fun),
+  receive
+    {ack, Ref} -> {ok, Proxy};
```
Comment (dszoboszlay) [inline src/vnet_proxy.erl:36]: You leave no longer needed monitor on the process here.

File: src/vnet_proxy.erl [hunk 12]
```diff
@@ -24,10 +23,19 @@
...
-                , init
-                , [VNodeC, VNodeS, Pid, Gen]
-                ).
+  P = self(),
+  Ref = make_ref(),
+  Fun =
+    fun() ->
+        init(VNodeC, VNodeS, Pid, Gen),
+        P ! {ack, Ref},
+        loop(Pid)
```
Comment (dszoboszlay) [inline src/vnet_proxy.erl:32]: I'd rather keep the ack-ing functionality in the `init/4` function.

If you stick to this structure though, you must update the type spec of `init/4` - it's no longer a `no_return()` function.

File: test/counter_server_example/test/counter_server_example.erl [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+%%%===================================================================
+
+%% @doc
+%% Simulating a scenario with one node as client, making a request.
+once_valid_scenario() ->
+  Nodes = [server, client],
+  {ok, S} = vnet:start_link(Nodes),
+  unlink(S),
+  %% Synchronous server setup
+  vnet:rpc(server, counter_server_example, setup_server, []),
```
Comment (dszoboszlay) [inline test/counter_server_example/test/counter_server_example.erl:169]: Use `?MODULE` instead of `counter_server_example` here and in later places too!

File: test/counter_server_example/test/counter_server_example.erl [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+%%% All _scenario functions demonstrate scenarios of using the
+%%% server. See their documentation for more info.
+%%%
+%%% All _no_vnet_scenario functions do not use vnet and their purpose
+%%% is to check the core implementation of the counter server.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(counter_server_example).
```
Comment (dszoboszlay) [inline test/counter_server_example/test/counter_server_example.erl:15]: I think the code in this module should be refactored a bit for readability.

My understanding is that the scenarios are actually permutations of 3 parameters:
1. How to call the server?
    * `no_vnet`: do not use `vnet` at all
    * `rpc`: use `vnet:rpc` calls
    * `proxy`: use `gen_server:call` with `{via, vnet, {server, ?SERVER}}`
2. What should the valid client do?
    * `single`: perform a single request (sometimes called `once` in the code)
    * `double`: perform two requests (sometimes called `twice` in the code)
    * `double_same_gen`: perform two requests assuming the generation is not incremented
3. What complication happens in the test concurrently?
    * no complications
    * `invalid`: an other client makes an invalid request
    * `disconnect`: the connection between the client and server is interrupted
    * `node_down`: the server node is killed

I would recommend consistently naming the scenarios according to the parameters they choose and using generic client/call/check etc. functions that cover all the possible cases and take these parameters as arguments.

File: test/counter_server_example/test/counter_server_example.erl [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+  ?assertEqual(G1, G2);
+check_same_gen(Res1, Res2) ->
+  ?assertEqual(ok, {unexpected, Res1, Res2}).
+
+%%%===================================================================
+
+%% @doc
+%% Simulating a scenario with one node as client, making a request.
+once_valid_scenario() ->
+  Nodes = [server, client],
```
Comment (dszoboszlay) [inline test/counter_server_example/test/counter_server_example.erl:165]: Maybe the name of the nodes should be defined as macros too for readability. E.g. `{via, vnet, {?server_node, ?server}}`.

**Review by dszoboszlay — CHANGES_REQUESTED** [review]

File: test/counter_server_example/test/counter_server_example.erl (hunk 14 above)
Comment (aronisstav) [reply test/counter_server_example/test/counter_server_example.erl:15]: I agree that a more consistent naming scheme can be established. However, these names are also used in the tables shown in the paper, which cannot be changed. Do you think it is acceptable to merge this as is and fix the names afterwards?

**Comment by aronisstav** [comment]
All comments are reasonable, but I'll wait for an answer to https://github.com/klarna/vnet/pull/1#discussion_r220833143 before working on fixes.

File: test/counter_server_example/test/counter_server_example.erl (hunk 14 above)
Comment (dszoboszlay) [reply test/counter_server_example/test/counter_server_example.erl:15]: Sure, leave them as they are.

Maybe even after the cleanup you could leave the original scenarios in the code like this:

```erlang
%% This XY paper refers to this scenario on its legacy name.
%% Please keep this wrapper for the reference of readers of the paper.
old_scenario_fashioned() ->
  new_fashioned_scenario().
```

**Comment by dszoboszlay** [comment]
Let's continue this in #4 !
